release-lib: build packages without `meta.platforms` on all platforms

emilazy opened the pull request:
This was added in <https://github.com/NixOS/nixpkgs/pull/19990> to ensure that packages that previously weren’t building at all would get at least one platform’s worth of coverage. However, since most `runCommand` calls don’t explicitly set any `meta` information, this meant that things like `tests.makeWrapper` were only being run for `x86_64-linux`. Since most trivial builder calls should work on all platforms, it doesn’t make much sense for them to be restricted in this way, and we shouldn’t unnecessarily penalize other platforms by treating an empty `meta.platforms` inconsistently with `check-meta`’s maximally‐permissive interpretation. Actual packages that only work on a subset of platforms should, of course, set `meta.platforms` explicitly.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy committed:
tests.build-environment-info: add more information

The user information would have helped diagnose the empty GECOS issue
we had with the Hydra Darwin builders recently and the OS information
seems generally nice to have.

emilazy committed:
release-lib: build packages without `meta.platforms` on all platforms

This was added in <https://github.com/NixOS/nixpkgs/pull/19990> to
ensure that packages that previously weren’t building at all would
get at least one platform’s worth of coverage. However, since most
`runCommand` calls don’t explicitly set any `meta` information,
this meant that things like `tests.makeWrapper` were only being
run for `x86_64-linux`. Since most trivial builder calls should
work on all platforms, it doesn’t make much sense for them to be
restricted in this way, and we shouldn’t unnecessarily penalize
other platforms by treating an empty `meta.platforms` inconsistently
with `check-meta`’s maximally‐permissive interpretation. Actual
packages that only work on a subset of platforms should, of course,
set `meta.platforms` explicitly.

emilazy commented:
(For posterity: I talked myself into the more radical version of this after seeing that the `[ "x86-64-linux" ]` default was not, as I had assumed, “we’re expanding platform support but old derivations without anything set will assume they only run on the default platform” but instead “these things aren’t building on Hydra at *all* and we should fix that”.)

philiptaron reviewed (approved):
Heck yes.

emilazy committed:
cudaPackages.cuda-samples: set `meta.platforms`

emilazy committed:
cudaPackages.cuda-library-samples: set `meta.platforms`

emilazy committed:
phpExtensions.*: use `meta.broken` rather than filtering

emilazy committed:
phpExtensions.*: inherit `meta.platforms` from PHP

RossComputerGuy reviewed (approved):
![i approb](https://github.com/user-attachments/assets/3ab33eae-df79-446b-9982-bc29233e9c3a)

emilazy commented:
Don’t get too excited, I’m still playing whack‐a‐mole with the evaluation failures. But hopefully we can get this in.

emilazy committed:
tests.nixos-functions: don’t omit derivations on non‐Linux

Hydra can handle packages that `throw` but not attributes that
disappear depending on platform, and it seems okay for evaluation to
fail here for non‐Linux platforms.

philiptaron reviewed (approved):
Master is now questionable, but I leave it up to the author about whether to retarget the PR to staging.

emilazy committed:
texlivePackages.*: set `meta.{platforms,hydraPlatforms}`

emilazy commented:
The latest push should be a net addition of about ~5k Hydra jobs by setting `hydraPlatforms` for TeX Live constituent packages, after discussion in the Staging room on Matrix. Some ~2.5k of the remaining jobs are actual useful SBCL and PHP packages that aren’t getting built for `aarch64-linux` or Darwin right now, and there are other, smaller package sets that will be built on other platforms where they weren’t before. All of this is good stuff that we ought to be building on supported platforms (given our decision to `recurseIntoAttrs` them in the first place), and that probably was never intended to be restricted to `x86_64-linux`.

There will also be jobs that don’t generate new builds, as discussed above. I think it should be okay as it is now, based on the discussion, and we can set more `meta.{platforms,hydraPlatforms}` if anything turns up that looks particularly problematic. I’d prefer for someone else to merge this, after seeing ofborg results (which will inflate the count a bit by not including the *reduction* in `texlivePackages.*`).

jtojnar commented on pkgs/top-level/php-packages.nix:
```diff
@@ -463,7 +463,7 @@ in {
           buildInputs = [ uwimap openssl pam pcre2 libkrb5 ];
           configureFlags = [ "--with-imap=${uwimap}" "--with-imap-ssl" "--with-kerberos" ];
           # Using version from PECL on new PHP versions.
-          enable = lib.versionOlder php.version "8.3";
```
I agree that the other instances should use `meta.broken` but this one specifically is needed to be filtered out on PHP > 8.3 since the PHP 8.4 source tree stopped shipping it and this broken derivation would shadow the fallback definition above.

emilazy replied:
Thanks for the insight. I’ve restored the logic using `lib.optionals` instead since it’s just this one package; let me know if you think that’s okay?

jtojnar replied:
Makes sense to me.

emilazy commented:
> Are we going to want to get this in before ZHF and before the merge?

I missed this comment. Yes, I think so, assuming we agree that it seems like a good idea to merge, although if it happens with the `staging-next` merge then it can also just target `master` since it makes no difference in terms of additional jobs.

emilazy commented:
ofborg results now list 9982 jobs, so minus the ~4445 TeX Live jobs that this is removing, it should be a net increase of 5.5k jobs, as expected. Should be ready when anyone feels brave enough.

RossComputerGuy commented:
> I missed this comment. Yes, I think so, assuming we agree that it seems like a good idea to merge, although if it happens with the `staging-next` merge then it can also just target `master` since it makes no difference in terms of additional jobs.

Sounds good, we probably should merge this in the next few hours or any time before 6PM PST to get the ball rolling. I hope to do the merge as listed in the 24.11 release schedule after work which would mean some time after 6PM PST.

emilazy commented:
Sure, I guess it could be merged now and another `staging-next` eval kicked off. Or it could be merged directly before the merge into `master`. I don’t know really; I don’t understand Hydra bottlenecks well enough to say what’s best and it was enough of a pain to figure out what was going on here and come up with a safe‐seeming plan to address it, so I’ll let someone else do the scary merge.

RossComputerGuy commented:
Guess I'll merge it.

emilazy commented:
Not sure whether to kick off a `staging-next` evaluation or not. There’s an installer test issue that is currently blocking the merge into `master`.

RossComputerGuy commented:
> Not sure whether to kick off a `staging-next` evaluation or not.

Idk how to kick one off but sure, why not? Better get some of that work rolling.

> There’s an installer test issue that is currently blocking the merge into `master`.

Is there an open issue? I'll look into that before merging staging-next.

emilazy commented:
Well, because there’ll presumably be a new eval to test the installer fix anyway. See https://github.com/NixOS/nixpkgs/pull/348827#issuecomment-2441081863. We cannot merge as it would block the channel.

SuperSandro2000 commented on pkgs/top-level/php-packages.nix:
```diff
@@ -655,7 +648,7 @@ in {
           configureFlags = [ "--with-snmp" ];
           # net-snmp doesn't build on darwin.
-          enable = (!stdenv.hostPlatform.isDarwin);
+          meta.broken = (!stdenv.hostPlatform.isDarwin);
```
Isn't this wrong now? Before it was enabled on linux and now it is marked broken on linux.

emilazy replied:
Yes, oops. But it turns out it builds fine on Darwin now. https://github.com/NixOS/nixpkgs/pull/353479

SuperSandro2000 replied:
I think pdo had the same mistake #353492
